src/data/structure/UserProfile.java

```diff
@@ -57,16 +57,21 @@ public String getId() {
      * @param preferenceId
      * @return added preference
      */
-    public Preference addPreference(PreferenceCategory category, String preferenceId) {
+    public Preference addPreference(PreferenceCategory category, String preferenceId, int popularity,
```

Comment:
My point is that no class outside of Preference should own the default values for a Preference's attributes. You should make a new constructor for Preference that only requires the original two parameters. Then, you can re-consolidate these two methods here.